src/main/java/org/radarcns/service/SourceService.java

```diff
@@ -78,16 +85,87 @@ public SourceService(SourceMonitorService sourceMonitorService, SourceCatalog so
                     "Cannot retrieve sourceType for given sourceType " + source.getSourceId());
         }
 
-        return new org.radarcns.domain.restapi.Source()
+        TimeFrame effectiveTimeFrame = this.sourceMonitorService
+                .getEffectiveTimeFrame(projectId, subjectId,
+                        source.getSourceId().toString(), sourceType);
+        return new Source()
                 .sourceId(source.getSourceId().toString())
                 .assigned(source.isAssigned())
                 .sourceName(source.getSourceName())
                 .sourceTypeCatalogVersion(source.getSourceTypeCatalogVersion())
                 .sourceTypeProducer(source.getSourceTypeProducer())
                 .sourceTypeModel(source.getSourceTypeModel())
-                .effectiveTimeFrame(this.sourceMonitorService
-                        .getEffectiveTimeFrame(projectId, subjectId,
-                                source.getSourceId().toString(), sourceType));
+                .effectiveTimeFrame(effectiveTimeFrame)
+                .status(getSourceStatus(source.getSourceTypeProducer(), source
+                        .getSourceTypeModel(), effectiveTimeFrame));
+    }
+
+    /**
+     * Build a {@link org.radarcns.domain.restapi.Source} using provided parameters and by
+     * calculating TimeFrame of the source.
+     *
+     * @param projectId of subject
+     * @param subjectId of subject
+     * @param source instance from MP
+     * @return computed Source.
+     */
+    private Source buildSource(String projectId, String subjectId,
+            SourceDTO source) {
+        SourceTypeDTO sourceType;
+        // a source fetched from MP should ideally have a source-type
+        try {
+            sourceType = this.sourceCatalog.getSourceType(source
+                    .getSourceType().getProducer(), source.getSourceType().getModel(), source
+                    .getSourceType().getCatalogVersion());
+        } catch (NotFoundException | IOException e) {
+            LOGGER.error(
+                    "Cannot retrieve sourceType for given sourceType " + source.getSourceId());
+            throw new IllegalStateException(
+                    "Cannot retrieve sourceType for given sourceType " + source.getSourceId());
+        }
+
+        TimeFrame effectiveTimeFrame = this.sourceMonitorService
+                .getEffectiveTimeFrame(projectId, subjectId,
+                        source.getSourceId(), sourceType);
+        return new Source()
+                .sourceId(source.getSourceId())
+                .assigned(source.getAssigned())
+                .sourceName(source.getSourceName())
+                .sourceTypeCatalogVersion(sourceType.getCatalogVersion())
+                .sourceTypeProducer(sourceType.getProducer())
+                .sourceTypeModel(sourceType.getModel())
+                .effectiveTimeFrame(effectiveTimeFrame)
+                .status(getSourceStatus(sourceType.getProducer(), sourceType.getModel(),
```

Comment:
Shouldn't this be in a Builder class like (SourceBuilder). I thought Source class is a POJO

Reply:
It indeed uses the builder pattern.